### PR TITLE
Add an ECS provider

### PR DESCRIPTION
**Pull request by lpetre**
I need to do a couple things, but I wanted to open this PR for discussion.

TODO:
 - [x] write tests
 - [x] update documentation
 - [x] update glide.yaml
 - [x] make sure provider.Watch is implemented correctly

### CODE REVIEW

**Comment by stingerpk** [comment]
+1

File: provider/ecs.go [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			select {
+			case <-stop:
+				cancel()
+			}
+		}()
+
+		operation := func() error {
+			select {
```
Comment (emilevauge) [inline provider/ecs.go]: Why did you add a select clause here?

File: provider/ecs.go [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+				}
+			}
+
+			return nil
+		}
+
+		notify := func(err error, time time.Duration) {
+			log.Errorf("ECS connection error %+v, retrying in %s", err, time)
+		}
+		err := backoff.RetryNotify(operation, job.NewBackOff(backoff.NewExponentialBackOff()), notify)
```
Comment (emilevauge) [inline provider/ecs.go]: Could you wrap it with `safe.OperationWithRecover(operation)`,  in case of a panic in the goroutine?

File: provider/ecs.go (hunk 1 above)
Comment (lpetre) [inline provider/ecs.go]: In the case where the stop channel has been sent a value, I cancel the context, my operation will return context.Canceled, and backoff.RetryNotify will call my operation again. So I handle that here.

**Comment by lpetre** [comment]
Thanks for the feedback @emilevauge. Regarding the goroutines in `provider/ecs.go`, there is one long running goroutine at [line 102](https://github.com/containous/traefik/pull/1088/files#diff-20ed89f06f1c428e51205096690cb934R102) that is responsible for watching the stop channel and cancelling the context, which I use to cancel any outstanding AWS API http requests. As far as I can tell this only can leak if `stop` is never signaled.

Otherwise there is a goroutine owned by the pool that is responsible for calling my `operation`. This will exit without error if the `ctx` context is cancelled.

I don't think there is a way for goroutines to leak here.

File: provider/ecs.go (hunk 1 above)
Comment (lpetre) [inline provider/ecs.go]: I'll push a fix, but without this select in the current form this code would retry forever when the stop channel is sent a value.  Because provider.loadECSConfig (called [here](https://github.com/containous/traefik/pull/1088/files#diff-20ed89f06f1c428e51205096690cb934R126) and [here](https://github.com/containous/traefik/pull/1088/files#diff-20ed89f06f1c428e51205096690cb934R140) will return the error `context.Canceled` (which is not-nil) and the backoff code will call the operation again, forever.

**Comment by SantoDE** [comment]
Wow @lpetre. Thanks a lot! Beside the little thing of @emilevauge , I'm good :)

File: provider/ecs.go (hunk 1 above)
Comment (lpetre) [inline provider/ecs.go]: I agree with your comment that the https://github.com/containous/traefik/pull/1088/files#diff-20ed89f06f1c428e51205096690cb934R149 also handles the case, but only if we're waiting for the refresh timer. If we had been in the middle of an AWS http request, that is when the select was useful.

**Comment by SantoDE** [comment]
Hey @lpetre ,

thanks for all your help! I'll review your latest commits soon. One thing upfront: There are conflicts in your glide files. Could you resolve them and rebase your branch to only have 1 commit? :)

**Comment by lpetre** [comment]
@SantoDE no problem, I'll do that as soon as I'm back at my computer.

**Comment by lpetre** [comment]
@SantoDE I've updated the glide files w/ the latest version in master. My glide.yaml changes weren't very big, but my glide.lock is very different than the version from this PR: https://github.com/containous/traefik/pull/1116

**Comment by lpetre** [comment]
This includes a glide change from https://github.com/containous/traefik/pull/1119

**Comment by lpetre** [comment]
@SantoDE this is rebased and up to date with latest master

**Comment by emilevauge** [comment]
@lpetre thanks a lot for you reactivity :clap:
